### ISSUE TITLE
create recipe for `<ComplexInput />`

### ISSUE THREAD

**Issue by malerba118** [chakra-ui/chakra-ui-docs#486]
### Description

Some complex inputs are hard to get right. For example number inputs are strangely hard to implement. I think there's a layer of abstraction that can mitigate the pain when implementing complex inputs that expect the user to enter a rigid syntax (eg json, css units,  csv, etc).

### Problem Statement/Justification

As mentioned, complex inputs can be hard to get right. Imagine an input field where the user can include a number with a variety of suffixes (10px, 10%, 10em, 10vw, etc). If the user types an invalid suffix, the value can't be accepted verbatim and the input must revert to a valid value or coerce the invalid value into a valid value.

### Proposed Solution or API

It is common practice for complex inputs to allow the user to type anything and then to process the value on blur, either accepting or rejecting it. If rejected, the value will typically revert to the last known valid value.

I think a `ComplexInput` component could be a good addition to chakra.

Strangely, i've never seen this level of abstraction in the wild, but i think it could make for a good building block for complex inputs that you might encounter in an app like figma (eg: number input, input with unit (%, em, rem, px), json input, etc).

[Here's a codesandbox with a loose implementation of the proposed api](https://codesandbox.io/s/complexinput-k9otd?file=/src/NumberInput.tsx).

Basically the idea is to pass custom parse/stringify implementations as props to an input component.

```jsx
 <ComplexInput
    stringify={stringify}
    parse={parse}
    value={value}
    onChange={onChange}
 />
```

When the input is blurred, the component will call parse and decide if the input is valid. 

If so, `onChange` is called with the valid, parsed input. 

However, if the user's input is invalid the component will revert to the last known valid value and `onChange` will not be called.
